### PR TITLE
fix os.isdir() and os.mkdir()

### PR DESCRIPTION
**Pull request by mercury233**
**What does this PR do?**

* On MinGW, `_wstat` does not fully recognize reparse points (symbolic links); `GetFileAttributesW` needs to be used instead.
* Fixed an issue where `os.mkdir` incorrectly returned success when the target already existed and was a file, added the test.
* Simplified the implementation of `os.isdir`.

**How does this PR change Premake's behavior?**

In the edge case where the target of `mkdir` already exists and is a file, the operation may now terminate earlier.

**Anything else we should know?**

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [ ] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [ ] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

*You can now [support Premake on our OpenCollective](https://opencollective.com/premake). Your contributions help us spend more time responding to requests like these!*

### CODE REVIEW

**Comment by mercury233** [comment]
The current test needs to be run outside the “sandbox” in order to pass, due to the reasons mentioned above.

* Run the test outside the sandbox.
* Skip this test on FreeBSD, or skip it only in the FreeBSD CI via an environment variable.
* Abandon this test completely.

I am not sure which option would be the most appropriate.

File: tests/base/test_os.lua [hunk 1]
```diff
@@ -179,6 +179,33 @@
...
+		test.istrue(os.rmdir(link))
+		test.istrue(os.rmdir(target))
+		test.istrue(os.rmdir(root))
+	end
+
+	function suite.mkdir_ReturnsError_OnExistingFile()
+		local ok, err = os.mkdir("folder/ok.lua")
+		test.isnil(ok)
+		test.isequal("string", type(err))
+		test.istrue(os.isfile("folder/ok.lua"))
```
Comment (Jarod42) [inline tests/base/test_os.lua]: I would move `test.istrue(os.isfile("folder/ok.lua"))` at the beginning. It is the prerequire for the failure of `os.mkdir`

**Review by Jarod42 — APPROVED** [review]

File: tests/base/test_os.lua [hunk 2]
```diff
@@ -179,6 +179,33 @@
 		test.isfalse(os.islink("folder/ok2.lua"))
 	end
 
+--
+-- os.mkdir() tests
+--
+
+	function suite.mkdir_ReturnsTrue_OnExistingDirectoryLink()
+		-- BSD CI mounts the workspace through SSHFS, so keep topology tests on local storage.
```
Comment (samsinsane) [inline tests/base/test_os.lua]: What was the actual problem?

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: tests/base/test_os.lua (hunk 2 above)
Comment (mercury233) [reply tests/base/test_os.lua]: https://github.com/premake/premake-core/actions/runs/30967252835/job/92183785538
after `os.linkdir("folder/subfolder", "folder/subfolder2")`, `os.mkdir("folder/subfolder2")` fail.

Comment (samsinsane) [reply tests/base/test_os.lua]: Why did it fail?

Comment (mercury233) [reply tests/base/test_os.lua]: I’m not sure. I can only guess that it has something to do with SSHFS.

Comment (nickclark2016) [reply tests/base/test_os.lua]: Likely related to issues that I started addressing with #2749, if I had to guess. Our BSD VM seems to be breaking quite a bit of code that works elsewhere. Not sure if it's with the VM set up or if it's BSD itself.

**Comment by nickclark2016** [comment]
I want to follow up on this. Currently, our CI jobs are passing. Is this a gap in our current testing?

**Comment by mercury233** [comment]
> I want to follow up on this. Currently, our CI jobs are passing. Is this a gap in our current testing?

The main issue is that previously, `os.mkdir` would return success when the target already existed as a file. Fixing this requires reliably checking the attributes of the target, which introduced symlink detection and, in turn, caused an issue in the FreeBSD CI.

~~The current test works around this issue, but it introduces file operations outside the sandbox (in a temporary directory).~~ I've dropped the newly added test, I think this probably wouldn’t cause any issues in practice.

File: tests/base/test_os.lua (hunk 2 above)
Comment (samsinsane) [reply tests/base/test_os.lua]: The comment reads as if it's obvious why you'd use local storage when using BSD and SSHFS, but then you're not sure why it doesn't work. We might just be calling the wrong functions on BSD? It would be nice to have more investigation into the issue so that we can clearly state why this test doesn't work with that configuration.

Comment (mercury233) [reply tests/base/test_os.lua]: > The comment reads as if it's obvious why you'd use local storage when using BSD and SSHFS, but then you're not sure why it doesn't work. We might just be calling the wrong functions on BSD? It would be nice to have more investigation into the issue so that we can clearly state why this test doesn't work with that configuration.

Sorry, I'm not interested about BSD. So I've just removed the test.

File: tests/base/test_os.lua [hunk 3]
```diff
@@ -179,6 +179,16 @@
 		test.isfalse(os.islink("folder/ok2.lua"))
 	end
 
+--
+-- os.mkdir() tests
+--
+
+	function suite.mkdir_ReturnsError_OnExistingFile()
+		test.istrue(os.isfile("folder/ok.lua"))
+		local ok, err = os.mkdir("folder/ok.lua")
```
Comment (nickclark2016) [inline tests/base/test_os.lua]: If this "breaks" the `ok.lua` file, this will cause subsequent tests depending on the existence of this file to fail. Let's create a temp file or temp directory to do this.

**Review by nickclark2016 — CHANGES_REQUESTED** [review]
One change requested in tests, but LGTM otherwise.

**Review by nickclark2016 — APPROVED** [review]

**Review by samsinsane — APPROVED** [review]
